### PR TITLE
Removing apache commons & Guava dependency 

### PR DESCRIPTION
**Pull request by vishwakarma**
This commit contains -- 
1. removal of apache commons dependency
2.  Writing own LCS since we no longer use apache commons lib.

### CODE REVIEW

**Comment by LouizFC** [comment]
This commit contains:

1. removal of guava library. Simple methods got inlined, others got moved to PathUtils.
2. creation of PathUtils:
*It serves as a low level internal api for dealing with low level paths/jsonpointers.
*pre-compiled patterns used when decoding, following https://github.com/flipkart-incubator/zjsonpatch/pull/54

**Comment by LouizFC** [comment]
Maybe we could further optimize the lcs algorithm. I am gonna take a look at this: https://github.com/fillumina/LCS

**Comment by vishwakarma** [comment]
Sure @LouizFC 
Regarding LCS, i don't think we could do better than O(N*M) but regardless of that I will have a look at the repo link which you shared.

**Comment by vishwakarma** [comment]
Hi @LouizFC 
So far --
1. All test-suite passed ( one checked-in with this library and other one i have which contains 16M operations)
2.  Earlier to run 16M operation test-suite, it used to take ~6 minutes 30 seconds, now it takes 5 minutes 57 seconds.

File: src/main/java/com/flipkart/zjsonpatch/InternalUtils.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        int size = input.size();
+        for (int i = 0; i < size; i++) {
+            toReturn.add(input.get(i));
+        }
+        return toReturn;
+    }
+
+    private static List<JsonNode> reverse(List<JsonNode> list) {
+        List<JsonNode> toReturn = new LinkedList<JsonNode>();
+        for (int i = list.size() - 1; i >= 0; i--) {
```
Comment (holograph) [inline src/main/java/com/flipkart/zjsonpatch/InternalUtils.java]: Would probably be better to forward-iterate the list (the interface makes no guarantees about time complexity on indexed retrieval/update) and build up an ArrayList (which does).

File: src/main/java/com/flipkart/zjsonpatch/JsonDiff.java [hunk 2]
```diff
@@ -50,16 +45,16 @@ public static JsonNode asJson(final JsonNode source, final JsonNode target, Enum
         generateDiffs(diffs, path, source, target);
 
         if (!flags.contains(DiffFlags.OMIT_MOVE_OPERATION)) {
-          /*
-         * Merging remove & add to move operation
-         */
+            /*
```
Comment (holograph) [inline src/main/java/com/flipkart/zjsonpatch/JsonDiff.java]: Why multi-line comments? Just makes everything more difficult...

File: src/main/java/com/flipkart/zjsonpatch/JsonDiff.java [hunk 3]
```diff
@@ -452,10 +441,6 @@ private static void compareObjects(List<Diff> diffs, List<Object> path, JsonNode
     }
 
     private static List<JsonNode> getLCS(final JsonNode first, final JsonNode second) {
-
-        Preconditions.checkArgument(first.isArray(), "LCS can only work on JSON arrays");
-        Preconditions.checkArgument(second.isArray(), "LCS can only work on JSON arrays");
-
-        return ListUtils.longestCommonSubsequence(Lists.newArrayList(first), Lists.newArrayList(second));
+        return InternalUtils.longestCommonSubsequence(InternalUtils.toList((ArrayNode) first), InternalUtils.toList((ArrayNode) second));
```
Comment (holograph) [inline src/main/java/com/flipkart/zjsonpatch/JsonDiff.java]: Slight change in error semantics, parameters no longer checked for whether or not they are arrays. Intentional?

File: src/main/java/com/flipkart/zjsonpatch/InternalUtils.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import com.fasterxml.jackson.databind.node.ArrayNode;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+class InternalUtils {
+
+    static List<JsonNode> toList(ArrayNode input) {
+        List<JsonNode> toReturn = new ArrayList<JsonNode>();
```
[the remark this answers is not in the document]
Comment (vishwakarma) [reply src/main/java/com/flipkart/zjsonpatch/InternalUtils.java]: Agreed

File: src/main/java/com/flipkart/zjsonpatch/InternalUtils.java (hunk 1 above)
Comment (vishwakarma) [reply src/main/java/com/flipkart/zjsonpatch/InternalUtils.java]: We could use Collections.reverse, which guarantees linear time complexity instead of this method

File: src/main/java/com/flipkart/zjsonpatch/JsonDiff.java (hunk 2 above)
Comment (vishwakarma) [reply src/main/java/com/flipkart/zjsonpatch/JsonDiff.java]: Agreed, will make it Single line comment

File: src/main/java/com/flipkart/zjsonpatch/JsonDiff.java (hunk 3 above)
Comment (vishwakarma) [reply src/main/java/com/flipkart/zjsonpatch/JsonDiff.java]: Since the caller already checks whether both inputs( source & target) are of Array type before calling this method, i thought to remove the additional checks.

File: src/main/java/com/flipkart/zjsonpatch/InternalUtils.java (hunk 1 above)
Comment (holograph) [reply src/main/java/com/flipkart/zjsonpatch/InternalUtils.java]: That requires a clone first, not sure it'll end up being any faster...

File: src/main/java/com/flipkart/zjsonpatch/JsonDiff.java (hunk 3 above)
Comment (holograph) [reply src/main/java/com/flipkart/zjsonpatch/JsonDiff.java]: Makes sense, but I checked and couldn't see the aforementioned check at the caller. Anyway it's only for clarity of errors when the API is misused, so not a critical point regardless.

File: src/main/java/com/flipkart/zjsonpatch/InternalUtils.java (hunk 1 above)
Comment (vishwakarma) [reply src/main/java/com/flipkart/zjsonpatch/InternalUtils.java]: Inside InternalUtils.longestCommonSubsequence, we could use Collections.reverse ( it does in-place reversal ) before the return statement. 
By your last comment, did you mean that sir?

File: src/main/java/com/flipkart/zjsonpatch/JsonDiff.java (hunk 3 above)
Comment (vishwakarma) [reply src/main/java/com/flipkart/zjsonpatch/JsonDiff.java]: Actually, caller of compareArray() methods checks the data-type of source & target.

File: src/main/java/com/flipkart/zjsonpatch/InternalUtils.java (hunk 1 above)
Comment (holograph) [reply src/main/java/com/flipkart/zjsonpatch/InternalUtils.java]: It's possible, but may be slightly less readable. Bottom line, we don't have a benchmark suite yet, so the whole discussion is kind of moot :-)

Comment (vishwakarma) [reply src/main/java/com/flipkart/zjsonpatch/InternalUtils.java]: I am using @cyber-front test-suite ( shared sometime back with both of us ) for benchmarking where it creates 65,536 random objects and performed a series of 256 updates on each.

File: src/main/java/com/flipkart/zjsonpatch/InPlaceApplyProcessor.java [hunk 5]
```diff
@@ -147,11 +147,11 @@ public void replace(List<String> path, JsonNode value) {
             if (isNullOrEmpty(fieldToReplace) && path.size() == 1)
                 target = value;
             else if (parentNode.isObject())
-                ((ObjectNode) parentNode).put(fieldToReplace, value);
+                ((ObjectNode) parentNode).set(fieldToReplace, value);
```
Comment (holograph) [inline src/main/java/com/flipkart/zjsonpatch/InPlaceApplyProcessor.java]: Why the change? Is this semantically equivalent?

Comment (LouizFC) [reply src/main/java/com/flipkart/zjsonpatch/InPlaceApplyProcessor.java]: The method is long deprecated in jackson api

Comment (LouizFC) [reply src/main/java/com/flipkart/zjsonpatch/InPlaceApplyProcessor.java]: Now that you commented, this change shouldn't be in this commit, it passed without me noticing, sorry. I will let you decide if this should pass or not

**Comment by vishwakarma** [comment]
@LouizFC 
Looks good, Thank you

**Comment by LouizFC** [comment]
You are welcome. Thank you too for letting me take part in this, it was a pleasure. I look forward for helping more in the future, if you want, of course.

**Comment by vishwakarma** [comment]
@LouizFC , Your are open to propose any change in this project or to even start a new project. I would be available.
How can i connect you over gmail or other social network for future?

**Comment by LouizFC** [comment]
@vishwakarma I have added my email in to my github profile. Thank you for your interest.
